native/src/utils/renderHtml.ts

```diff
@@ -33,72 +52,174 @@ const renderJS = (cacheDictionary: Record<string, string>) => `
     }
   })();
 
-  (function() {
-    var container = document.getElementById('measure-container')
+  (function addWebviewHeightListeners() {
+    const container = document.getElementById('measure-container')
 
-    function adjustHeight () {
-      container.setAttribute('style', 'padding: 1px 0;'); // Used for measuring collapsed vertical margins
+    function adjustHeight() {
+      container.setAttribute('style', 'padding: 1px 0;') // Used for measuring collapsed vertical margins
 
       if (!window.ReactNativeWebView) {
-        return window.setTimeout(adjustHeight, 100);
+        return window.setTimeout(adjustHeight, 100)
       }
 
-      var height = container.getBoundingClientRect().height - 2
-      window.ReactNativeWebView.postMessage(JSON.stringify({ type: '${HEIGHT_MESSAGE_TYPE}', height: height }));
-      container.setAttribute('style', '');
+      const height = container.getBoundingClientRect().height - 2
+      window.ReactNativeWebView.postMessage(JSON.stringify({ type: '${HEIGHT_MESSAGE_TYPE}', height: height }))
+      container.setAttribute('style', '')
     }
 
-    window.addEventListener('load', adjustHeight);
-    window.addEventListener('resize', adjustHeight);
-    var details = document.querySelectorAll("details")
-    details.forEach(detail => detail.addEventListener("toggle", adjustHeight))
+    window.addEventListener('load', adjustHeight)
+    window.addEventListener('resize', adjustHeight)
+    const details = document.querySelectorAll('details')
+    details.forEach(detail => detail.addEventListener('toggle', adjustHeight))
   })();
+
+  (function handleIframes() {
+    function showMessage(text, element, className, iframeSource) {
+      const textNode = document.createTextNode(text)
+      if (className) {
+        element.classList.add(className)
+      }
+
+      if (iframeSource) {
+        element.appendChild(document.createTextNode(iframeSource))
+        element.appendChild(document.createElement('br'))
+      }
+      element.appendChild(textNode)
+    }
+
+    function showSettingsButton(element) {
+      function onClickHandler() {
+        window.ReactNativeWebView.postMessage(JSON.stringify({ type: '${SETTINGS_MESSAGE_TYPE}' }))
+      }
+
+      const buttonLabel = '${t('layout:settings')}'
+      const button = document.createElement('button')
+      button.name = 'opt-in-settings-button'
+      button.innerHTML = buttonLabel
+      button.id = button.name
+      button.onclick = onClickHandler
+      element.appendChild(button)
+    }
+
+    function showMessageWithSettings(text, iframeContainer, iframeSource) {
+      showMessage(text, iframeContainer, 'iframe-info-text', iframeSource)
+      showSettingsButton(iframeContainer)
+    }
+
+    function showOptIn(text, iframeContainer, source) {
+      function onClickHandler() {
+        window.ReactNativeWebView.postMessage(
+          JSON.stringify({ type: '${IFRAME_MESSAGE_TYPE}', allowedSource: { type: source, allowed: true } }),
```

Comment:
Optional: Since `allowed` is only ever set to `true`, we could change this to:
```suggestion
          JSON.stringify({ type: '${ALLOW_EXTERNAL_SOURCE_MESSAGE_TYPE}', source }),
```